### ISSUE TITLE
Emulation doesnt boot every time in CP/M while selecting menu point 3

### ISSUE THREAD

**Issue by guidol70** [GmEsoft/Z80-MBC2_VGA32#3]
First: Man Thanks for this emulation. Its great to have this aside the RunCPM on a TTGO VGA32!
;)

I compiled your emulator and can - without problems - set the (german) keyboard, other videomode, Timezone and WiFi.

Mostly on the every first boot after powering on I can boot into CP/M 2.2 (normal Diskset) or CP/M 3.0 (Diskset 2).

But after using the emulation and repowering I often can select the menu-option 3 (after diskset for CP/M 3.0) and then the emulation will halt/freeze after showing that the Z80 will run now.

I tested it with 3 different SD-Cards, because I first thought it could be a SD-Card problem,
but I do get this with a 16GB & 32GB SanDisk and a 32GB Lexar Card (all formatted to FAT32 with the "SD Card Formatter"-tool)

Do you got an idea if that if a timing problem? Or is my SDCard to "big"?

Is there any reboot or reset option - other than repowering? The reset-button near the "user-button" IO36 doenst work for me :(

BTW: When unset the keyboard-value does show "-1" which isnt in the list ;)

**Comment by GmEsoft** [comment]
Hi Guido,

Thanks for your appreciation.

I mostly use CP/M 3.0 on my board (TTGO VGA32 v1.4) without encountering any problem. I use 2 different MicroSD cards Sandisk Ultra 16GB and 32GB. I did not reformat them...

I'm not aware of any possible timing problems with the CP/M 3.0 BIOS implementation...

The reset and user buttons work both fine for me, and I use the reset button to reboot the VGA32... perhaps your device is faulty... ? 

The uninitialized keyboard value shows -1 by design, when not set in Preferences, so the default keyboard mapping can be set automatically to US layout (value 3).

Note: I just rebuilt the sketch using FabGL v1.0.3 from FDiVitto and esp32 from Espressif Systems v1.0.6 installed using this board manager URL: https://raw.githubusercontent.com/espressif/arduino-esp32/gh-pages/package_esp32_dev_index.json .

And I connect a genuine PS/2 keyboard to the keyboard socket, and no mouse.

**Comment by guidol70** [comment]
> The reset and user buttons work both fine for me, and I use the reset button to reboot the VGA32... perhaps your device is faulty... ?
> Note: I just rebuilt the sketch using FabGL v1.0.3 from FDiVitto and esp32 from Espressif Systems v1.0.6 installed 

Hi,
I also use FabGL v1.0.3 and Espressif v1.0.6
Reset-Button is now working - dont know why it wasnt working in the first place.
I also got a real PS/2-keyboard (no USB-dual-variant) and no mouse connected.
I checked again another PSU and didnt see any difference.

The other thing which I observed is that it do work better when I wait a minute before repowering the VGA32.

For me it looks like you use the WiFi only for a few seconds to get the time and then disconnect, because I cant see VGA32 with a IP on my net.

**Comment by GmEsoft** [comment]
> For me it looks like you use the WiFi only for a few seconds to get the time and then disconnect, because I cant see VGA32 with a IP on my net.

Indeed. The device connects to the Wi-Fi to get the time, it then sets the clock and then does a warm boot without reconnecting to Wi-Fi. This is because the Wi-Fi activation consumes too much internal SRAM, and so limits the amount of SRAM available to the VGA graphics mode in higher resolutions, and to the Z80-MBC2 emulation itself.

**Comment by Unixware** [comment]
Same issues like  guidol70  : CP/M 3 freezes, reset works but the device stuck (no video) so I unplug the power. We may have different devices (different components you see, typical chinesium crap) 
my specs:  TTGO VGA32 1.4 - datecode 21-2-20

**Comment by Unixware** [comment]
Just an update,... @guidol70 you are right... issues are because of the uSD cards... all of them in my case were this shite brand: Transcend.. ! I ve change them (one 8GB the others 16GB)  with some TDKs and Samsungs all ok!
